Add language to gear menu for logged out users

amanagr opened the pull request:
Fixes #21961

![image](https://user-images.githubusercontent.com/25124304/166212743-da77b2e0-bb38-457d-9de5-4ed94bda1ebe.png)

amanagr commented on static/js/settings_display.js:
```diff
@@ -43,33 +45,49 @@ function change_display_setting(data, $status_el, success_msg_html, sticky) {
+
+                const $link = $(e.target).closest("a[data-code]");
+                Cookies.set(page_params.language_cookie_name, $link.attr("data-code"));
+                window.location.reload();
```
I am not sure how reasonable doing a reload here is vs telling user to reload it.

amanagr commented:
@alya this is ready for your review.

andersk commented on zerver/lib/home.py:
```diff
@@ -175,9 +175,13 @@ def build_page_params_for_home_page_load(
 
+    default_language = register_ret["user_settings"]["default_language"]
+    if user_profile is None:
+        default_language = request.COOKIES.get(settings.LANGUAGE_COOKIE_NAME, default_language)
```
I think there’s no need to call `get_and_set_request_language` in this case. If the correct language is stored in Django’s cookie, it will already have been set up by Django’s `LocaleMiddleware` and stored in `request.LANGUAGE_CODE`.

amanagr replied:
Makes sense.

amanagr commented on static/js/admin.js:
```diff
@@ -3,7 +3,7 @@ import $ from "jquery";
 import render_admin_tab from "../templates/settings/admin_tab.hbs";
```
![image](https://user-images.githubusercontent.com/25124304/166413030-3cfa197f-3358-4c18-b7e0-492d65e2db2c.png)

amanagr commented on frontend_tests/puppeteer_tests/admin.ts:
```diff
@@ -312,25 +312,6 @@ async function test_organization_profile(page: Page): Promise<void> {
     await page.waitForSelector(gravatar_selctor, {visible: true});
 }
 
-async function submit_default_user_settings(page: Page): Promise<void> {
```
Since the code is already tested in `change_language` in settings.ts of puppeteer tests, I don't think we need to redo the tests here.

amanagr commented:
@sahil839 it would be great if you can review this too.

sahil839 commented:
Do we not want to use "save-discard" widget for this setting? I think it seems weird to let one setting in the section behave differently. From what I tried tomorrow, it should not be much complex to do it if we want.

amanagr commented:
> Do we not want to use "save-discard" widget for the `Notification language` setting? I think it seems weird to let one setting in the section behave differently. From what I tried tomorrow, it should not be much complex to do it if we want.

Oh, yeah! We can just use save discard widget here.

amanagr commented:
On second thought! Anything more than just a `Saved` message is an inconvenience here since it is hard to select a wrong language in this case.
![image](https://user-images.githubusercontent.com/25124304/166456309-8c9f1666-6ff6-4796-87d5-c92eed53c483.png)

sahil839 commented:
I feel many of the other realm settings are also dropdown, and kinda similar to this setting which also contains a list of lanugages to choose from. Plus I feel would be weird to have just one setting not use save-discard widget. Also when I change one setting in that same section and then change the language setting, the indicator moves to right with save and discard button still shown.

amanagr commented:
> Also when I change one setting in that same section and then change the language setting, the indicator moves to right with save and discard button still shown.

I fixed this issue.

> Plus I feel would be weird to have just one setting not use save-discard widget. 

Let's see what others feel like. I am happy to make changes for the best!

timabbott commented:
Thanks for doing this!  I did a quick rebase. I think the `fa-language` icon is weird; a globe (the circle with latitudes, not the one we're using for web-public streams) is the traditional icon for this. We can try importing the icon we previously tried for web-public streams for as the `language` Zulip icon.

![image](https://user-images.githubusercontent.com/2746074/166748090-e20e8362-4e42-4a3e-8969-e5aa9ba7029f.png)

* I think for the spectator view at least, we want to change the heading to "Select language" rather than "Select default language". It's probably the case we should just change that heading unconditionally, since "Select default language" is worse in all 3 places we show this modal :).
* This popup appears after changing the language. I think we probably shouldn't try to turn it off -- it likely only appears if the language is one your browser is not configured to think you can read, and in any case, that widget is useful if you're looking at conversations in a language you don't know. I verified, that it could translate a Zulip message in Spanish into English for me ... though it only popped up after I selected Spanish as the UI language. (The right-click "Translate" menu in Chrome works too). So That's a nice feature in general, and perhaps one worth mentioning as a side comment in our documentation somewhere. @alya FYI. 
![image](https://user-images.githubusercontent.com/2746074/166748814-180a161f-c4c0-4c3e-bffa-9e6702379466.png)
* For the organization settings UI, we should have the label appear on its own line like we do for adjacent settings. See  ![image](https://user-images.githubusercontent.com/2746074/166750083-c7ed6b64-bb02-4a42-bf5f-fcb63387be66.png). I suppose we should also probably restyle these to not use blue text, since we don't use that in other buttons.
* I agree with Sahil that we should use a save/discard widget for this input -- that's a pattern we use in 100% of organization settings prior to this change, and it's best to be consistent about that sort of thing.

I think my main question is whether we want to use the cookie mechanism to switch the language, or if we want to instead just redirect you to, e.g., /de/. I suppose the cookie is better because random links you get from elsewhere will not be to the /de/ URLs, so year, let's go with the cookie.

amanagr commented:
![image](https://user-images.githubusercontent.com/25124304/167331465-c7ead56d-7543-4c9a-941d-e467e10bb131.png)
![image](https://user-images.githubusercontent.com/25124304/167331492-5414399d-4cd6-42ca-b092-944374049688.png)
![image](https://user-images.githubusercontent.com/25124304/167331534-58a11516-6071-46cd-937f-c9f569c37d39.png)

@timabbott addressed all your comments. @sahil839 can you please do another review too?

sahil839 commented on static/js/settings_display.js:
```diff
@@ -90,7 +90,7 @@ function default_language_modal_post_render($container) {
-export function launch_default_language_setting_modal($container) {
+export function launch_default_language_setting_modal() {
     const html_body = render_dialog_default_language({
         language_list: get_language_list_columns(user_settings.default_language),
```
This seems to be a bug, we would need to pass `realm_default_language` for notification language modal.

amanagr replied:
I am not sure what you mean. Why do we need to pass it here? The handlebar doesn't seem to be using it.

sahil839 replied:
I mean that we should pass `realm_default_language` to `get_language_list_column` for notification language modal because we show the selected language in different color in the modal. You can reproduce it by changing notification lamguage to be something other than the user setting language and then open the modal again.

amanagr replied:
Oh! good point.

sahil839 reviewed (commented):
Posted a couple of comments. Also, while testing the PR, I noticed that the discard button does not work as expected. Clicking on discard should change the language name in the button to the original language.

amanagr commented on static/js/settings_display.js:
```diff
@@ -57,6 +64,24 @@ function default_language_modal_post_render() {
+                $language_element.text(new_language);
+                $language_element.attr("data-language-code", setting_value);
+                settings_org.save_discard_widget_status_handler($("#org-notifications"));
+            });
```
Yeah.

alya commented:
Hm, I don't know if there's something wrong with how I'm loading it, but the icon seems to be missing from the menu for me.

![Screen Shot 2022-05-26 at 3 50 21 PM](https://user-images.githubusercontent.com/2090066/170592295-ef5c0cab-2c08-47d0-864b-a1d3b8ddea2e.png)

alya commented:
I agree with all the comments above.

While we're here, would it be easy to increase the padding so that the columns in the language selection menu aren't so tightly squashed together?  I can file a separate issue if that would be better.

![Screen Shot 2022-05-26 at 3 54 49 PM](https://user-images.githubusercontent.com/2090066/170592608-61400397-5af8-4bd0-9e6a-073c91df0427.png)

Thanks!

amanagr commented:
Added more vertical padding and addressed all of @sahil839 's comments.
<img width="1143" alt="Screenshot 2022-05-30 at 5 46 44 PM" src="https://user-images.githubusercontent.com/25124304/170990744-1eb256b0-8189-478e-be89-f0e4fdc0611f.png">


@alya you can try hard reloading to fix icon issues.

sahil839 commented on static/styles/settings.css:
```diff
@@ -569,9 +573,28 @@ input[type="checkbox"] {
+
+.realm_default_language {
+    .input-group {
+        display: inline;
```
This CSS is actually not applied since there is no `input-group` element inside `realm_default_language` element.

amanagr replied:
Looks like I forgot to remove the css I added for the other direction was heading with this.

sahil839 commented on static/styles/settings.css:
```diff
@@ -569,9 +573,28 @@ input[type="checkbox"] {
+    }
+
+    .help_link_widget_fake_label {
+        display: inline-block;
```
Not sure if we actually use this CSS. Cannot find such element.

amanagr replied:
Yup.

sahil839 reviewed (commented):
Posted some comments. Also, the "pencil" icon and the language name are too close in the button. Previously, there was some space between them.

amanagr commented on static/js/settings_display.js:
```diff
@@ -41,7 +48,7 @@ function change_display_setting(data, $status_el, success_msg_html, sticky) {
         $status_el.data("sticky_msg_html", success_msg_html);
     }
-    settings_ui.do_settings_change(channel.patch, "/json/settings", data, $status_el, opts);
+    settings_ui.do_settings_change(channel.patch, patch_url, data, $status_el, opts);
```
Yeah.

amanagr commented on static/js/settings_org.js:
```diff
@@ -501,6 +501,16 @@ function discard_property_element_changes(elem, for_realm_default_settings) {
+            $("#org-notifications .language_selection_widget .language_selection_button span").attr(
+                "data-language-name",
+                get_language_name(property_value),
+            );
```
Yeah, nice catch.

amanagr commented on static/styles/settings.css:
```diff
@@ -569,9 +573,28 @@ input[type="checkbox"] {
-    text-decoration: none;
-    margin-left: 20px;
+.language_selection_widget {
+    margin-bottom: 20px;
```
I removed it because, it wasn't working well here.

sahil839 replied:
Ok, for it to work with botht the widgets, we need to add the class to `.language_selection_widget` element, but other than that I am not sure why this wouldn't work since it applies exact same CSS.

amanagr replied:
Ok, I changed it, since I don't recall the reason why I removed it. It may have to do with `input-group` being associated with dropdowns.

amanagr commented on static/styles/settings.css:
```diff
@@ -569,9 +573,28 @@ input[type="checkbox"] {
+    margin-bottom: 20px;
+
+    .title {
+        margin-bottom: 5px;
```
Yeah.

amanagr commented:
@sahil839 thanks for the great review! Updated the PR.

sahil839 commented:
Posted one comment above about using `input-group`. Rest looks LGTM

alya commented:
Thanks for the updates -- looks good to me!

alya commented:
@timabbott this has been reviewed by @sahil839 and myself -- over to you. :)

timabbott commented:
Rebased and merged, with this small tweak:

```
diff --git a/static/styles/settings.css b/static/styles/settings.css
index af9a0de272..54be51f25f 100644
--- a/static/styles/settings.css
+++ b/static/styles/settings.css
@@ -278,7 +278,7 @@ td .button {
     }
 
     td {
-        padding-top: 3px;
+        padding-top: 1px;
     }
 }
 
```

I made this change because with 3px padding, we end up with the widget being too small and having an unnecessary scrollbar:

![image](https://user-images.githubusercontent.com/2746074/171521268-031c4648-cdca-4b1f-b749-8d370819f025.png)

The right fix is probably to move to 3 columns (remembering to test the responsive mobile experience) -- let's do that in a follow-up PR. 

Huge thanks for doing this everyone! I'm very glad to have this cleaned up.
